// @colyseus/redis-driver/build/experimental_luaScripts/RedisDriver.mjs
// packages/drivers/redis-driver/src/experimental_luaScripts/RedisDriver.ts
import {
  Redis,
  Cluster
} from "ioredis";
import {
  debugMatchMaking,
  initializeRoomCache
} from "@colyseus/core";
import { FILTER_AND_SORT_SCRIPT, FIND_ONE_SCRIPT } from "./luaScripts.mjs";
var ROOMCACHES_KEY = "roomcaches";
var RedisDriver = class {
  constructor(options, clusterOptions) {
    // Cache concurrent filterAndSort requests to avoid redundant Redis calls
    this._pendingFilterRequests = /* @__PURE__ */ new Map();
    this._pendingFindOneRequests = /* @__PURE__ */ new Map();
    this._client = Array.isArray(options) ? new Cluster(options, clusterOptions) : new Redis(options);
  }
  async has(roomId) {
    return await this._client.hexists(ROOMCACHES_KEY, roomId) === 1;
  }
  async query(conditions, sortOptions) {
    const conditionsJson = JSON.stringify(conditions || {});
    const sortOptionsJson = sortOptions ? JSON.stringify(sortOptions) : "";
    const cacheKey = `${conditionsJson}:${sortOptionsJson}`;
    let pendingRequest = this._pendingFilterRequests.get(cacheKey);
    if (!pendingRequest) {
      pendingRequest = this._client.filterAndSort(
        ROOMCACHES_KEY,
        conditionsJson,
        sortOptionsJson
      );
      this._pendingFilterRequests.set(cacheKey, pendingRequest);
      pendingRequest.finally(() => {
        this._pendingFilterRequests.delete(cacheKey);
      });
    }
    const results = await pendingRequest;
    return results.map((roomcache) => initializeRoomCache(JSON.parse(roomcache)));
  }
  async cleanup(processId) {
    const cachedRooms = await this.query({ processId });
    debugMatchMaking("removing stale rooms by processId %s (%s rooms found)", processId, cachedRooms.length);
    const itemsPerCommand = 500;
    for (let i = 0; i < cachedRooms.length; i += itemsPerCommand) {
      await this._client.hdel(ROOMCACHES_KEY, ...cachedRooms.slice(i, i + itemsPerCommand).map((room) => room.roomId));
    }
  }
  async findOne(conditions, sortOptions) {
    if (typeof conditions.roomId !== "undefined") {
      const roomcache = await this._client.hget(ROOMCACHES_KEY, conditions.roomId);
      if (roomcache) {
        return initializeRoomCache(JSON.parse(roomcache));
      }
      return void 0;
    } else {
      const conditionsJson = JSON.stringify(conditions || {});
      const sortOptionsJson = sortOptions ? JSON.stringify(sortOptions) : "";
      const cacheKey = `${conditionsJson}:${sortOptionsJson}`;
      let pendingRequest = this._pendingFindOneRequests.get(cacheKey);
      if (!pendingRequest) {
        pendingRequest = this._client.findOneRoom(
          ROOMCACHES_KEY,
          conditionsJson,
          sortOptionsJson
        );
        this._pendingFindOneRequests.set(cacheKey, pendingRequest);
        pendingRequest.finally(() => {
          this._pendingFindOneRequests.delete(cacheKey);
        });
      }
      const result = await pendingRequest;
      if (result) {
        return initializeRoomCache(JSON.parse(result));
      }
      return void 0;
    }
  }
  async update(room, operations) {
    if (operations.$set) {
      for (const field in operations.$set) {
        if (operations.$set.hasOwnProperty(field)) {
          room[field] = operations.$set[field];
        }
      }
    }
    if (operations.$inc) {
      for (const field in operations.$inc) {
        if (operations.$inc.hasOwnProperty(field)) {
          room[field] += operations.$inc[field];
        }
      }
    }
    await this._client.hset(ROOMCACHES_KEY, room.roomId, JSON.stringify(room));
    return true;
  }
  async persist(room, _ = false) {
    if (!room.roomId) {
      debugMatchMaking("RedisDriver: can't .persist() without a `roomId`");
      return false;
    }
    await this._client.hset(ROOMCACHES_KEY, room.roomId, JSON.stringify(room));
    return true;
  }
  async remove(roomId) {
    const result = await this._client.hdel(ROOMCACHES_KEY, roomId);
    return result > 0;
  }
  async shutdown() {
    await this._client.quit();
  }
  async boot() {
    this._client.defineCommand("filterAndSort", {
      numberOfKeys: 1,
      lua: FILTER_AND_SORT_SCRIPT
    });
    this._client.defineCommand("findOneRoom", {
      numberOfKeys: 1,
      lua: FIND_ONE_SCRIPT
    });
  }
  //
  // only relevant for the test-suite.
  // not used during runtime.
  //
  clear() {
    this._client.del(ROOMCACHES_KEY);
  }
};
export {
  RedisDriver
};
